scheduler/server/task_scheduler.go

```diff
@@ -81,48 +82,54 @@ func (s *statefulScheduler) getTaskAssignments() []taskAssignment {
 	return assignments
 }
 
+type nodeStatesByNodeID map[cluster.NodeId]*nodeState
+
 // Helper fn, appends to 'assignments' and updates nodeGroups.
 // Should successfully assign all given tasks if caller invokes this with self-consistent params.
-func assign(
-	cs *clusterState,
-	tasks []*taskState,
-	stat stats.StatsReceiver,
-) (assignments []taskAssignment) {
-	idleNodeGroupIDs := map[string]string{}
-	for groupID, group := range cs.nodeGroups {
+// Note: there may be a race condition between recognizing idle nodes as available for assignment and
+// nodes becoming offlined or suspended. The code does the best it can, but it may assign a task to
+// a node that clusterState considers offlined/suspended before or as the task is actually being started
+func (s *statefulScheduler) assign(tasks []*taskState) (assignments []taskAssignment) {
+	idleNodesByGroupIDs := map[string]nodeStatesByNodeID{}
+
+	// make a local copy of (non-suspended/offlined) idle nodeIDs by groupID, these will be the nodes assigned to the tasks
+	for groupID, group := range s.clusterState.nodeGroups {
 		if len(group.idle) > 0 {
-			idleNodeGroupIDs[groupID] = groupID
+			idleNodesByGroupIDs[groupID] = nodeStatesByNodeID{}
+			for _, ns := range group.idle {
+				if !ns.suspended() && !s.clusterState.isOfflined(ns) {
+					idleNodesByGroupIDs[groupID][ns.node.Id()] = ns
+				}
+			}
 		}
 	}
+
 	for _, task := range tasks {
-		var idleNodeGroupID string
 		var nodeSt *nodeState
 
 		// is there a node group (with idle node) for this snapshot?
-		if _, ok := cs.nodeGroups[task.Def.SnapshotID]; ok {
-			nodeSt = findIdleNodeInGroup(cs, task.Def.SnapshotID)
+		if nodeGroup, ok := idleNodesByGroupIDs[task.Def.SnapshotID]; ok {
+			nodeSt = s.findIdleNodeInGroup(nodeGroup)
 		}
-		if nodeSt != nil {
-			idleNodeGroupID = task.Def.SnapshotID
-		} else {
+		if nodeSt == nil {
 			// could not find any free nodes in node group for the task's snapshot id.  Look for a free node in the other
 			// node groups, starting with the "" node group
-			if _, ok := idleNodeGroupIDs[""]; ok {
-				nodeSt = findIdleNodeInGroup(cs, "")
+			if nodeGroup, ok := idleNodesByGroupIDs[""]; ok {
+				nodeSt = s.findIdleNodeInGroup(nodeGroup)
 			}
 			if nodeSt == nil {
```

Comment:
we're already in an if block for if nodeSt == nil here, we don't need to check again

Reply:
ln 118 may set nodeSt.

Reply:
whoops!